api/src/main/java/org/xwiki/apps/googleapps/CookieAuthenticationPersistenceStoreTools.java

```diff
@@ -19,7 +19,261 @@
  */
 package org.xwiki.apps.googleapps;
 
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Cipher;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.SecretKeySpec;
+import javax.servlet.http.Cookie;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang3.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xwiki.stability.Unstable;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.web.XWikiRequest;
+
+/** Copied code from xwiki-authenticator-trusted
+ * https://github.com/xwiki-contrib/xwiki-authenticator-trusted/edit/master\
+ *   /xwiki-authenticator-trusted-api/src/main/java/org/xwiki/contrib/authentication\
+ *   /internal/CookieAuthenticationPersistenceStore.java.
+ * @version $Id$
+ * @since 2.5-RC1
+ */
 public class CookieAuthenticationPersistenceStoreTools
 {
+    private static final String AUTHENTICATION_CONFIG_PREFIX = "xwiki.authentication";
+
+    private static final String COOKIE_PREFIX_PROPERTY = AUTHENTICATION_CONFIG_PREFIX + ".cookieprefix";
+    private static final String COOKIE_PATH_PROPERTY =  AUTHENTICATION_CONFIG_PREFIX + ".cookiepath";
+    private static final String COOKIE_DOMAINS_PROPERTY = AUTHENTICATION_CONFIG_PREFIX + ".cookiedomains";
+    private static final String ENCRYPTION_KEY_PROPERTY = AUTHENTICATION_CONFIG_PREFIX + ".encryptionKey";
+
+    private static final String CIPHER_ALGORITHM = "TripleDES";
+
+    private static final String AUTHENTICATION_COOKIE = "XWIKITRUSTEDAUTH";
+
+    /**
+     * The string used to prefix cookie domain to conform to RFC 2109.
+     */
+    private static final String COOKIE_DOT_PFX = ".";
+
+    private static final String EQUAL_SIGN = "=";
+    private static final String UNDERSCORE = "_";
+
+    /* Default Cookie TTL : 3 days */
+    private static final int COOKIE_TTL = 259200;
+
+    private Logger logger = LoggerFactory.getLogger(CookieAuthenticationPersistenceStoreTools.class);
+
+    private XWikiContext context;
```

Comment:
Better to inject components and even better to use a provider and get the component only when it's needed:
```
@Inject
private Provider<XWikiContext> xwikiContextProvider;
```

Reply:
Mmmh. The class CookieAuthenticationPersistenceStoreTools is request-local thus far. So making it a component means that we change and generalize this to a singleton. Right? I attempted to not change the logic during porting but I acknowledge this makes sense.
paul

Reply:
> So making it a component means that we change and generalize this to a singleton. Right?

Not necessarily. See https://extensions.xwiki.org/xwiki/bin/view/Extension/Component%20Module#HUsingAnnotations . If you don't want you component to be ``@Singleton`` you can annotate it as

    @InstantiationStrategy(ComponentInstantiationStrategy.PER_LOOKUP)

Reply:
Thanks for the hints. All done.

Reply:
This took me a while but I am through at the last commit.